### PR TITLE
Launch efa instances from heterogenous reservations

### PR DESCRIPTION
**Pull request by arjkesh**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description

Update EFA tests to use a heterogenous approach for launching instances from capacity reservation. This allows us to make use of reservations where there may be < min number of instances available in a particular reservation, but there are enough instances available in multiple reservations or in reservations + open capacity

### Tests run

- Successfully launch p5 instances ([3100f27](https://github.com/aws/deep-learning-containers/pull/3768/commits/3100f277e69ec05f04008208189d31df1b428424))
- Successfully launch with p4d instances ([9bb1796](https://github.com/aws/deep-learning-containers/pull/3768/commits/9bb1796673069ee0d372d0438c93c8eb1f83ff3b))
- Test launching from one ([cbb3eab](https://github.com/aws/deep-learning-containers/pull/3768/commits/cbb3eab94505950e1772847eed723c937f0ad23a))
- ...and then launching quickly again to see failure mode ([01f5ee6](https://github.com/aws/deep-learning-containers/pull/3768/commits/01f5ee6719b500e6ac67d6547ba97c4c24049bd4))
- test default PR behavior ([8731694](https://github.com/aws/deep-learning-containers/pull/3768/commits/87316944f874ca3566b7609fa90f558a5a9eee51))
- additional sanity test for deepcopy update - [d904077](https://github.com/aws/deep-learning-containers/pull/3768/commits/d9040778ca75b4ec7ab3838ad3109dcba7b3e705)

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [ ] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**
<details>
<summary>Expand</summary>

- [ ] `sagemaker_remote_tests = true`
- [ ] `sagemaker_efa_tests = true`
- [ ] `sagemaker_rc_tests = true`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

</details>

### Formatting
- [ ] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

#### Builds to Execute
<details>
<summary>Expand</summary>

Click the checkbox to enable a build to execute upon merge.

*Note: By default, pipelines are set to "latest". Replace with major.minor framework version if you do not want "latest".*

- [ ] build_pytorch_training_latest
- [ ] build_pytorch_inference_latest
- [ ] build_tensorflow_training_latest
- [ ] build_tensorflow_inference_latest

</details>

### PR Checklist 
<details>
<summary>Expand</summary>

- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

#### NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ec2_benchmark_tests = true` or `sagemaker_benchmark_tests = true`
</details>

### Pytest Marker Checklist
<details>
<summary>Expand</summary>

- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type
</details>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/test_utils/ec2.py [hunk 1]
```diff
@@ -406,6 +408,150 @@ def launch_instances_with_retry(
...
+            if reservation["AvailabilityZone"] == most_common_az
+        ]
+
+        available_instances = sum(
+            [reservation["AvailableInstanceCount"] for reservation in reservations]
+        )
+
+        instances = []
+        try:
+            while available_instances and len(instances) < minimum_number_of_instances:
```
Comment (roywei) [inline test/test_utils/ec2.py:500]: if instance launch takes about 10 min, will `available_instances` change in the meantime?

File: test/test_utils/ec2.py [hunk 2]
```diff
@@ -406,6 +408,150 @@ def launch_instances_with_retry(
...
+            f"Launched too many instances somehow, raising and cleaning up - {instances}; min/max_allowed = {minimum_number_of_instances}"
+        )
+    return False
+
+
+class HeterogenousReservationError(Exception):
+    pass
+
+
+def launch_efa_with_heterogenous_reservations(
```
Comment (roywei) [inline test/test_utils/ec2.py:470]: if this works, we can deprecate `launch_efa_with_reservations`, I can't think of a use case where 2 instances must come from same cr

File: test/test_utils/ec2.py (hunk 1 above)
Comment (arjkesh) [reply test/test_utils/ec2.py:500]: yeah good point, will see if it is straightforward to recalculate this every time we try to launch an instance

File: test/test_utils/ec2.py [hunk 3]
```diff
@@ -406,6 +408,150 @@ def launch_instances_with_retry(
...
+    ec2_client, ec2_instance_type, ec2_run_instances_definition, fn_name=""
+):
+    minimum_number_of_instances = ec2_run_instances_definition["MinCount"]
```
[the remark this answers is not in the document]
Comment (arjkesh) [reply test/test_utils/ec2.py:508]: updated fn description

File: test/test_utils/ec2.py (hunk 1 above)
Comment (arjkesh) [reply test/test_utils/ec2.py:541]: refreshing reservations and available instances in every az loop and every launch try

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_pytorchddp_inductor.py [hunk 4]
```diff
@@ -35,7 +37,10 @@ def can_run_pytorchddp(ecr_image):
     return Version(image_framework_version) in SpecifierSet(">=1.10")
 
 
-# Skip due to known issue: https://github.com/pytorch/pytorch/issues/99074
+@pytest.mark.skipif(
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_pytorchddp_inductor.py:40]: I can remove these conditional skips if we still want to run these

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_pytorchddp.py [hunk 5]
```diff
@@ -35,6 +37,10 @@ def can_run_pytorchddp(ecr_image):
     return Version(image_framework_version) in SpecifierSet(">=1.10")
 
 
+@pytest.mark.skipif(
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_pytorchddp.py:40]: I can remove these conditional skips if we still want to run these

**Review by roywei — APPROVED** [review]

**Review by Lokiiiiii — APPROVED** [review]

**Pull request merged**